go/apps/mixnet/proxy.go

```diff
@@ -72,51 +83,99 @@ func (p *ProxyContext) SendMessage(c net.Conn, msg []byte) (int, error) {
 
 	bytes := copy(cell[1+n:], msg)
 	if _, err := c.Write(cell); err != nil {
-		return 0, err
+		return err
 	}
 
 	for bytes < msgBytes {
 		zeroCell(cell)
-		bytes += copy(cell, msg[bytes:])
+		cell[0] = msgCell
+		bytes += copy(cell[1:], msg[bytes:])
 		if _, err := c.Write(cell); err != nil {
-			return bytes, err
+			return err
 		}
 	}
-
-	return bytes, nil
+	return nil
 }
 
-// ReceiveMessage waits for a reply or error message from the router.
-func (p *ProxyContext) ReceiveMessage(c net.Conn, msg []byte) (int, error) {
+// ReceiveMessage reads message cells from the router and assembles them into
+// a messsage.
+func (p *ProxyContext) ReceiveMessage(c *Conn) ([]byte, error) {
 	var err error
+
+	// Receive cells from router.
 	cell := make([]byte, CellBytes)
 	if _, err = c.Read(cell); err != nil && err != io.EOF {
-		return 0, err
+		return nil, err
 	}
 
-	if cell[0] == msgCell { // Read a message.
-		// TODO(cjpatton)
-
-	} else if cell[0] == dirCell { // Handle a directive.
-		dirBytes, n := binary.Uvarint(cell[1:])
+	if cell[0] == dirCell {
 		var d Directive
-		if err := proto.Unmarshal(cell[1+n:1+n+int(dirBytes)], &d); err != nil {
-			return 0, err
+		err = unmarshalDirective(cell, &d)
+		if err != nil {
+			return nil, err
+		} else if *d.Type == DirectiveType_ERROR {
+			return nil, errors.New("router error: " + (*d.Error))
 		}
+		return nil, errCellType
+	} else if cell[0] != msgCell {
+		return nil, errCellType
+	}
+
+	msgBytes, n := binary.Uvarint(cell[1:])
+	if msgBytes > MaxMsgBytes {
+		return nil, errMsgLength
+	}
+
+	msg := make([]byte, msgBytes)
+	bytes := copy(msg, cell[1+n:])
 
-		switch *d.Type {
-		case DirectiveType_ERROR:
-			return 0, errors.New("router error: " + (*d.Error))
-		case DirectiveType_FATAL:
-			return 0, errors.New("router error: " + (*d.Error) + " (connection closed)")
-		default:
-			return 0, errBadDirective
+	for err != io.EOF && uint64(bytes) < msgBytes {
```

Comment:
I think there was a max message size before; is that being enforced here, or are we counting on the correctness of the mixnet server implementation? Otherwise, this could grow to be an unreasonable size.